File: python/Problem38.py
# Problem: 38. Count and Say
# URL: https://leetcode.com/problems/count-and-say/
# Difficulty: Medium
# Tags: String

class Solution:
  def countAndSay(self, n: int) -> str:
    say: str = "1"
    for i in range(1, n):
      newSay: str = ""
      prevDigit: chr = say[0]
      count: int = 1
      for j in range(1, len(say)):
        if prevDigit == say[j]:
          count += 1
        else:
          newSay += str(count) + str(prevDigit)
          prevDigit = say[j]
          count = 1
      newSay += str(count) + str(prevDigit)
      say = newSay
    return say

File: cpp/Problem38.cpp
/*
 * Problem: 38. Count and Say
 * URL: https://leetcode.com/problems/count-and-say/
 * Difficulty: Medium
 * Tags: String
 */
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
	string countAndSay(int n) {
		vector<char> say = {'1'};
		vector<char> newSay;
		for (int i = 1; i < n; i++) {
			newSay.clear();
			char prevDigit = say[0];
			int count = 1;
			for (int j = 1, size = say.size(); j < size; j++) {
				if (prevDigit == say[j]) {
					count++;
				} else {
					newSay.push_back(count + '0');
					newSay.push_back(prevDigit);
					prevDigit = say[j];
					count = 1;
				}
			}
			newSay.push_back(count + '0');
			newSay.push_back(prevDigit);
			say = newSay;
		}
		return string(say.begin(), say.end());
	}
};